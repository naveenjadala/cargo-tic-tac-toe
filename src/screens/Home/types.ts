type Matrix = string[][];
type playerProps = 'X' | 'O';
export type {Matrix, playerProps};
