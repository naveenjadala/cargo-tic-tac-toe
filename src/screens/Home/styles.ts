import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  container: {
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  text: {
    fontSize: 20,
    fontWeight: 'bold',
    marginVertical: 20,
  },
});
