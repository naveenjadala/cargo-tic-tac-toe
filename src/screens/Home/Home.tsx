import React, {useEffect, useState} from 'react';
import {ScrollView, Text, View} from 'react-native';

import {styles} from './styles';
import Board from '../../components/board/Board';
import Button from '../../components/button/Button';
import WinnerComponent from '../../components/winnerComponent/WinnerComponent';
import {Matrix, playerProps} from './types';
import {winningPatterns} from '../../assets/constants';

/**
 * Function that renders the Home component.
 *
 * @return {JSX.Element} The Home component
 */
const Home = () => {
  let matrix: Matrix = [
    ['', '', ''],
    ['', '', ''],
    ['', '', ''],
  ];

  const [board, setBoard] = useState<Matrix>(matrix);
  const [player, setPlayer] = useState<playerProps>('X');
  const [win, setWin] = useState<boolean>(false);
  const [draw, setDraw] = useState(false);
  const isInitialRender = React.useRef(true);

  useEffect(() => {
    if (!isInitialRender.current) {
      checkIsWinner();
    } else {
      isInitialRender.current = false;
    }
  }, [board]);

  /**
   * Updates the board with the player's move if the value is empty and the board
   * is updated.
   *
   * @function onPlay
   *
   * @param {number} i - the row index of the value
   * @param {number} j - the column index of the value
   * @return {void}
   */
  const onPlay = (i: number, j: number) => {
    if (board[i][j] === '' && !win) {
      setBoard(prev => {
        const newBoard = [...prev];
        newBoard[i][j] = player;
        return newBoard;
      });
    }
  };

  /**
   * Resets the game by setting the board to the initial state,
   * setting the player to 'X', and setting the win status to false.
   */
  const resetGame = () => {
    isInitialRender.current = true;
    setWin(false);
    setDraw(false);
    setPlayer('X');
    setBoard(matrix);
  };

  /**
   * Check if every element in the row is equal to the first element and not empty.
   *
   * @param {string[]} row - the array of strings to be checked
   * @return {boolean} true if every element meets the condition, false otherwise
   */
  const checkNonEmptyRow = (row: string[]) => {
    return row.every((x, i, arr) => x === arr[0] && x !== '');
  };

  /**
   * Function to check for a winner and update game state accordingly.
   */
  const checkIsWinner = () => {
    for (const pattern of winningPatterns) {
      if (checkNonEmptyRow(pattern.map(([i, j]) => board[i][j]))) {
        setWin(true);
        return;
      }
    }

    const isDraw = board.every(row => row.every(value => value !== ''));
    if (isDraw) {
      setDraw(true);
    } else {
      setPlayer(prev => (prev === 'X' ? 'O' : 'X'));
    }
  };

  return (
    <View style={styles.container}>
      <Text style={styles.text}>Players: {player}'s turn to play</Text>
      <Board matrix={board} board={board} onPlay={onPlay} />
      <Button resetGame={resetGame} label={'Reset Game'} />
      {win ? (
        <View testID="winner">
          <WinnerComponent
            label="winner"
            player={player}
            resetGame={resetGame}
          />
        </View>
      ) : null}
      {draw ? (
        <View testID="draw">
          <WinnerComponent
            label={'draw'}
            player={player}
            resetGame={resetGame}
          />
        </View>
      ) : null}
    </View>
  );
};

export default Home;
