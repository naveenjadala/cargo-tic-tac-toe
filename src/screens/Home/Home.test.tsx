import {fireEvent, render, screen, waitFor} from '../../test/test-utils';

import Home from './Home';

describe('Home', () => {
  test('Should render snapshot', () => {
    render(<Home />);
    expect(screen).toMatchSnapshot();
  });

  test('Should render first player is x', () => {
    render(<Home />);
    const text = screen.getByText("Players: X's turn to play");
    expect(text).toBeTruthy();
  });

  test('Should render second player is o', async () => {
    render(<Home />);
    fireEvent.press(screen.getByTestId('tile-00'));
    const text = screen.getByText("Players: O's turn to play");
    await waitFor(() => {
      expect(text).toBeTruthy();
    });
  });

  test('Should render correct player on press', async () => {
    render(<Home />);

    const textPlayerX = screen.getByText("Players: X's turn to play");
    expect(textPlayerX).toBeTruthy();

    fireEvent.press(screen.getByTestId('tile-00'));

    await waitFor(() => {
      expect(screen.getAllByTestId('tile')[0].props.children).toBe('X');
    });

    const textPlayerO = screen.getByText("Players: O's turn to play");
    expect(textPlayerO).toBeTruthy();

    fireEvent.press(screen.getByTestId('tile-01'));

    await waitFor(() => {
      expect(screen.getAllByTestId('tile')[1].props.children).toBe('O');
    });

    expect(textPlayerX).toBeTruthy();
  });

  test('should restart game', () => {
    render(<Home />);
    fireEvent.press(screen.getByTestId('tile-00'));
    const textPlayerO = screen.getByText("Players: O's turn to play");

    expect(textPlayerO).toBeTruthy();

    expect(screen.getAllByTestId('tile')[0].props.children).toBe('X');

    fireEvent.press(screen.getByTestId('resetGame'));

    const textPlayerX = screen.getByText("Players: X's turn to play");
    expect(textPlayerX).toBeTruthy();

    expect(screen.getAllByTestId('tile')[0].props.children).toBe('');
  });

  test('Should render x as winner', () => {
    render(<Home />);
    fireEvent.press(screen.getByTestId('tile-00')); // x
    fireEvent.press(screen.getByTestId('tile-01')); // o
    fireEvent.press(screen.getByTestId('tile-11')); // x
    fireEvent.press(screen.getByTestId('tile-10')); // o
    fireEvent.press(screen.getByTestId('tile-22')); // x
    const textPlayerX = screen.getByText("Players: X's turn to play");
    expect(textPlayerX).toBeTruthy();
    const text = screen.getByTestId('winner');
    waitFor(() => {
      expect(text).toBeTruthy();
    });
  });

  test('Should render o as winner', () => {
    render(<Home />);
    fireEvent.press(screen.getByTestId('tile-11')); // x
    fireEvent.press(screen.getByTestId('tile-00')); // o
    fireEvent.press(screen.getByTestId('tile-22')); // x
    fireEvent.press(screen.getByTestId('tile-01')); // o
    fireEvent.press(screen.getByTestId('tile-20')); // x
    fireEvent.press(screen.getByTestId('tile-02')); // o
    const textPlayerO = screen.getByText("Players: O's turn to play");
    expect(textPlayerO).toBeTruthy();
    const text = screen.getByTestId('winner');
    waitFor(() => {
      expect(text).toBeTruthy();
    });
  });

  test('Should render a draw game', () => {
    render(<Home />);
    fireEvent.press(screen.getByTestId('tile-11')); // x
    fireEvent.press(screen.getByTestId('tile-00')); // o
    fireEvent.press(screen.getByTestId('tile-22')); // x
    fireEvent.press(screen.getByTestId('tile-01')); // o
    fireEvent.press(screen.getByTestId('tile-02')); // x
    fireEvent.press(screen.getByTestId('tile-12')); // o
    fireEvent.press(screen.getByTestId('tile-10')); // x
    fireEvent.press(screen.getByTestId('tile-20')); // o
    fireEvent.press(screen.getByTestId('tile-21')); // x
    const text = screen.getByTestId('draw');
    waitFor(() => {
      expect(text).toBeTruthy();
    });
  });
});
