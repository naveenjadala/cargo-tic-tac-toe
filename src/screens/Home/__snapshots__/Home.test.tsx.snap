// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Should render snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
      "minHeight": "100%",
      "minWidth": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "marginVertical": 20,
        }
      }
    >
      Players: 
      X
      's turn to play
    </Text>
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "width": "80%",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-00"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-01"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-02"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "width": "80%",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-10"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-11"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-12"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "width": "80%",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-20"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-21"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderRadius": 10,
              "height": 100,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 100,
            }
          }
          testID="tile-22"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
              }
            }
            testID="tile"
          />
        </View>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignSelf": "center",
          "backgroundColor": "#8C8C8C",
          "borderRadius": 10,
          "height": 50,
          "justifyContent": "center",
          "margin": 5,
          "marginVertical": 20,
          "opacity": 1,
          "width": "80%",
        }
      }
      testID="resetGame"
    >
      <Text
        style={
          {
            "color": "#fff",
            "fontSize": 20,
            "textAlign": "center",
            "textAlignVertical": "center",
          }
        }
      >
        Reset Game
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
